src/lib/utils.ts

```diff
@@ -282,23 +283,55 @@ export function encodeKey(
     case 'mappingwithgrouping':
     case 'singleton':
     case 'mapping':
+      if (schema.valueType.includes('[CompactBytesArray]')) {
+        const valueType = schema.valueType.replace('[CompactBytesArray]', '');
+        const valueContent = schema.valueContent.replace(
+          '[CompactBytesArray]',
+          '',
+        );
+
+        if (
+          Array.isArray(value) &&
+          Array.isArray(value[0]) &&
+          valueType[0] === '(' &&
+          valueType[valueType.length - 1] === ')'
+        ) {
+          const encodedTuples = value.map((element) => {
+            return encodeTupleKeyValue(valueContent, valueType, element);
+          });
+          return encodeValueType('bytes[CompactBytesArray]', encodedTuples);
+        }
+        if (Array.isArray(value) && !Array.isArray(value[0]))
+          return encodeValueType(schema.valueType, value as string[]);
+      }
+
       if (isValidTuple(schema.valueType, schema.valueContent)) {
-        if (!Array.isArray(value)) {
+        if (
+          !Array.isArray(value) ||
+          (Array.isArray(value) && Array.isArray(value[0]))
+        ) {
           throw new Error(
             `Incorrect value for tuple. Got: ${value}, expected array.`,
           );
         }
-        return encodeTupleKeyValue(
-          schema.valueContent,
-          schema.valueType,
-          value,
-        );
+        if (Array.isArray(value) && !Array.isArray(value[0]))
+          return encodeTupleKeyValue(
+            schema.valueContent,
+            schema.valueType,
+            value as string[] | number[] | JSONURLDataToEncode[],
+          );
       }
 
       return encodeKeyValue(
         schema.valueContent,
         schema.valueType,
-        value,
+        value as
```

Comment:
can we assert value is not string[][] before this rather than have this big type cast

Reply:
I am not sure how we can do that

Reply:
Interesting so, doing it with 2 ifs works:

<img width="778" alt="image" src="https://user-images.githubusercontent.com/477945/216851548-ae2fe353-9ce4-49bc-a93d-6198ae77931f.png">

but with one if it doesnt work:

<img width="852" alt="image" src="https://user-images.githubusercontent.com/477945/216851581-83605b29-98b4-466e-8a98-975873810258.png">

I did a few checks tlocally and could manage to make it work.

Let's keep it as it is.